Date components - limit min date fiscal year feature (2785)

sfount opened the pull request:
This PR implements the functionality described in https://github.com/IMA-WorldHealth/bhima-2.X/issues/2785. 

The flag `limit-min-fiscal` has been applied to the `bh-date-interval` and `bh-date-editor` components, these are the two primarily used for reports. 

This flag will limit the date editor so that it can only select up to the beginning of the enterprises first fiscal year. 

This PR applies the `limit-min-fiscal` flag to the following modules: 
* Cash report 
* Cashflow by service 
* Client report 
* Cashflow 
* Stock exit 
* Balance sheet 
* Stock inventory

Dates in 2015, this is the earliest fiscal year so these are valid
![screenshot from 2018-09-02 19-51-53](https://user-images.githubusercontent.com/2844572/44959645-c0334a80-aee9-11e8-8b6e-0c0cf10222f7.png)

Dates in 2014 are before the opening fiscal year, these are invalid
![screenshot from 2018-09-02 19-52-06](https://user-images.githubusercontent.com/2844572/44959652-d3461a80-aee9-11e8-84f5-a9d562b98a21.png)

Year view 
![screenshot from 2018-09-02 19-52-25](https://user-images.githubusercontent.com/2844572/44959655-d6d9a180-aee9-11e8-9d95-b5596c2e2065.png)

Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/2785.

sfount committed:
feat(reports): component fiscal date limit

This commit introduces an optional flag on the `bhDateInterval`
component that is used on many reports. It will request the starting
date of the first fiscal year from the server and ensure the date input
cannot select a date before this.

sfount committed:
chore(reports) apply limit-min-fiscal to reports

This commit applies the `limit-min-fiscal` flag to all reports using the
`bh-date-interval`, ensuring users cannot provide dates that are before
financial records began.

sfount committed:
chore(reports): apply limit to bh-date-editor

This commit applied the `limit-min-fiscal` flag the `bh-date-editor`
component. It also updates reports that use this component to ensure
they cannot be submitted for before the enterprise fiscal activity.

jniles commented on client/src/js/components/bhDateEditor.js:
```diff
@@ -12,10 +12,11 @@ angular.module('bhima.components')
       disabled : '<?',
       dateFormat : '@?',
       label : '@?',
+      limitMinFiscal : '@?',
```
Hey, this is awesome!  Never thought it would be this easy!

jniles commented on client/src/js/components/bhDateEditor.js:
```diff
@@ -24,6 +25,11 @@ bhDateEditorController.$inject = ['bhConstants'];
+ * An optional limit-min-fiscal flag can be provided that will limit the date
+ * selection to after the start of enterprise financial records - note that
+ * limit-min-fiscal will *not* override min-date, it will only be applied if
+ * min-date has not been set.
```
Nice comments 👍

jniles reviewed (approved):
Nice!  This has been a feature we've wanted for a long time.  Code looks good to me.

Could you write an integration test for the new route you added?  After that, this PR could be merged.

sfount committed:
test(reports): tests for limit fiscal year feature

This commit introudces integration tests for the
`enterprises/:id/fiscal_start` route. It also udpates the
`bh-date-editor` unit tests to verify the `limit-min-fiscal` is correctly
parsed.

jniles reviewed (approved):
Fantastic feature.  Code and behavior look good to me!

bors r+

jniles committed:
chore(bhDateInterval): rename pickerOptions

This commit renames `pickerOptions` to `pickerFromOptions` to match its
`pickerToOptions` counterpart.
